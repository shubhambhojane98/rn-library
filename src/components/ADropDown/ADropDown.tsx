import React, {useState} from 'react';
import {View, Text, TouchableOpacity, FlatList, StyleSheet} from 'react-native';
import {moderateScale} from 'react-native-size-matters';
import ATextInput from '../ATextInput/ATextInput';
import {Color} from '../../theme';
import {defaultScale} from '../../utils/Common';
import AModal from '../AModal/AModal';

interface Prop {
  data: Array<any>;
  placeholder: string;
  height?: number;
  borderColor?: string;
  disable?: boolean;
  errorText?: string;
  marginBottom?: number;
  marginLeft?: number;
  marginRight?: number;
  marginTop?: number;
  renderItemCustom?: (_: any) => void;
}

const defaultDropDownHeight = 120;
const defaultMargin = 0;

const ADropDown: React.FC<Prop> = ({
  data,
  height = defaultDropDownHeight,
  borderColor,
  placeholder,
  disable,
  errorText,
  marginBottom = defaultMargin,
  marginLeft = defaultMargin,
  marginRight = defaultMargin,
  marginTop = defaultMargin,
  renderItemCustom,
}) => {
  const toggleRef = React.useRef<any>();

  const [selectedItem, setSelectedItem] = useState<any>();
  const [showOption, setShowOption] = useState(false);

  const [dropDownWidth, setDropDownWidth] = React.useState(undefined);
  const [dropDownTop, setDropDownTop] = React.useState<number | undefined>(
    undefined,
  );
  const [dropDownLeft, setDropDownLeft] = React.useState<number | undefined>(
    undefined,
  );
  const [icon, setIcon] = useState('downarrow');
  const calculatedPadding = renderItemCustom ? 0 : 12;

  const stylesWithProps = styles({
    dropDownTop,
    dropDownLeft,
    dropDownWidth,
    height,
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    calculatedPadding,
  });

  const onSelect = (item: any) => {
    setSelectedItem(item);
    setShowOption(false);
  };

  const renderItem = (item: any) => (
    <TouchableOpacity
      onPress={() => {
        setIcon('downarrow');
        onSelect(item);
      }}>
      {renderItemCustom ? (
        <>{renderItemCustom(item)}</>
      ) : (
        <Text style={stylesWithProps.dropDownItem}>{item.value}</Text>
      )}
    </TouchableOpacity>
  );

  const toggleDropdown = () => {
    setIcon('close');
    setShowOption(!showOption);
    toggleRef.current.measure(
      (_: any, __: any, width: any, ___: any, px: any, py: any) => {
        if (py > 640) {
          setDropDownWidth(width);
          setDropDownTop(py - 150);
          setDropDownLeft(px - 20);
        } else {
          setDropDownWidth(width);
          setDropDownTop(py + 35);
          setDropDownLeft(px - 20);
        }
      },
    );
  };

  return (
    <View
      ref={toggleRef}
      renderToHardwareTextureAndroid
      style={stylesWithProps.containerStyle}>
      <View>
        <ATextInput
          placeholder={placeholder}
          editable={false}
          value={selectedItem ? selectedItem.value : ''}
          disable={disable}
          errorText={errorText}
          rightIcon={icon}
          borderColor={borderColor}
          marginBottom={errorText ? 5 : 0}
          marginLeft={0}
          marginRight={0}
          marginTop={0}
        />
        <TouchableOpacity
          activeOpacity={1}
          onPress={toggleDropdown}
          style={stylesWithProps.dropDownButton}
        />
      </View>
      {!disable && dropDownTop && (
        <AModal
          backgroundColor="transparent"
          transparentOverlay={true}
          animationType="fade"
          visible={showOption}>
          <TouchableOpacity
            activeOpacity={1}
            onPress={() => {
              setShowOption(false);
              setIcon('downarrow');
            }}
            style={[
              stylesWithProps.dropDownContent,
              {...stylesWithProps.dropDownBackdrop},
            ]}>
            <View
              style={[
                stylesWithProps.dropDownContent,
                stylesWithProps.dropDownBackground,
              ]}>
              <FlatList
                data={data}
                renderItem={({item}) => renderItem(item)}
                keyExtractor={item => item.id}
                showsVerticalScrollIndicator
                nestedScrollEnabled={true}
              />
            </View>
          </TouchableOpacity>
        </AModal>
      )}
    </View>
  );
};

const styles = (props: {
  dropDownTop?: number;
  dropDownLeft?: any;
  dropDownWidth?: any;
  height?: number;
  marginTop?: number;
  marginRight?: number;
  marginBottom?: number;
  marginLeft?: number;
  calculatedPadding: number;
}) =>
  StyleSheet.create({
    containerStyle: {
      marginTop: props.marginTop,
      marginRight: props.marginRight,
      marginBottom: props.marginBottom,
      marginLeft: props.marginLeft,
    },
    dropDownContent: {
      padding: moderateScale(props.calculatedPadding, defaultScale),
      justifyContent: 'center',
      shadowColor: Color.black,
      shadowOffset: {
        width: 0,
        height: 0,
      },
      shadowOpacity: 0.2,
      shadowRadius: 3,
      elevation: 10,
    },
    dropDownBackdrop: {
      position: 'absolute',
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
    },
    dropDownBackground: {
      backgroundColor: Color.white,
      position: 'absolute',
      top: props.dropDownTop,
      left: props.dropDownLeft,
      width: props.dropDownWidth,
      height: moderateScale(
        props.height || defaultDropDownHeight,
        defaultScale,
      ),
    },
    dropDownItem: {
      textAlign: 'left',
      height: moderateScale(30, defaultScale),
      justifyContent: 'center',
    },
    dropDownButton: {
      position: 'absolute',
      top: 0,
      left: 0,
      width: '100%',
      height: '100%',
    },
  });

export default ADropDown;
