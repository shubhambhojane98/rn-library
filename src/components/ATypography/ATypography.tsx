import React from 'react';
import {StyleProp, StyleSheet, Text, TextProps, TextStyle} from 'react-native';
import {moderateScale} from 'react-native-size-matters';
import {fontFamilies} from '../../constants/fontFamilies';
import {TextAlignment, TypographyVariant} from './ATypographyEnum';
import {defaultScale} from '../../utils/Common';
import {withTheme, useTheme} from '../../core/theming';
import type {Theme} from '../../utils/types';
interface OwnProps {
  variant?: TypographyVariant;
  color?: string;
  style?: StyleProp<TextStyle>;
  textAlign?: TextAlignment;
  fontSize?: number;
  theme: Theme;
}

const defaultFontSize = moderateScale(14, defaultScale);

type Props = OwnProps & TextProps;
const Typography: React.FC<Props> = ({
  variant = TypographyVariant.PRIMARY,
  color,
  style,
  textAlign = TextAlignment.Left,
  children,
  fontSize = defaultFontSize,
  ...otherProps
}) => {
  const {colors} = useTheme();
  return (
    <Text
      style={[
        styles({
          fontSize,
        })[variant],
        {color: color ? color : colors.textColor},
        {textAlign},
        style,
      ]}
      {...otherProps}>
      {children}
    </Text>
  );
};

export default withTheme(Typography);

const styles = (props: {fontSize: number}) =>
  StyleSheet.create({
    primary: {
      fontFamily: `${fontFamilies.Fonts.primary}-Regular`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'normal',
    },
    primarySemiBold: {
      fontFamily: `${fontFamilies.Fonts.primary}-SemiBold`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: '600',
    },
    primaryBold: {
      fontFamily: `${fontFamilies.Fonts.primary}-Bold`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'bold',
    },
    secondary: {
      fontFamily: `${fontFamilies.Fonts.secondary}-Regular`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'normal',
    },
    secondaryDemi: {
      fontFamily: `${fontFamilies.Fonts.secondary}-Demi`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: '600',
    },
    secondaryBold: {
      fontFamily: `${fontFamilies.Fonts.secondary}-Bold`,
      fontSize: moderateScale(props.fontSize, defaultScale),
      fontWeight: 'bold',
    },
  });
