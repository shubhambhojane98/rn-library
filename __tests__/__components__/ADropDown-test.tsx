import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import ADropDown from '../../src/components/ADropDown/ADropDown';
import ATypography from '../../src/components/ATypography/ATypography';
import {TypographyVariant} from '../../src/components/ATypography/ATypographyEnum';

const user = [
  {id: 2, title: 'ritesh', value: 'ritesh'},
  {id: 3, title: 'amit', value: 'amit'},
  {id: 4, title: 'rohan', value: 'rohan'},
  {id: 6, title: 'virat', value: 'virat'},
  {id: 7, title: 'rohit', value: 'rohit'},
  {id: 8, title: 'rahul', value: 'rahul'},
];

it('renders correctly with default props', () => {
  const tree = renderer
    .create(<ADropDown placeholder="placeholder" data={user} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with custom render', () => {
  const tree = renderer
    .create(
      <ADropDown
        placeholder="placeholder"
        data={user}
        renderItemCustom={item => {
          return (
            <ATypography
              variant={TypographyVariant.PRIMARY_BOLD}
              style={{
                padding: 20,
              }}>
              {item.title}
            </ATypography>
          );
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with disabled & custom height', () => {
  const tree = renderer
    .create(
      <ADropDown
        placeholder="placeholder"
        data={user}
        disable={true}
        height={180}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
