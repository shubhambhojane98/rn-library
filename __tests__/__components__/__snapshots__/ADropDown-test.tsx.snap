// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with custom render 1`] = `
<View
  renderToHardwareTextureAndroid={true}
  style={
    Object {
      "marginBottom": 0,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 0,
    }
  }
>
  <View>
    <View>
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFF",
                "borderRadius": 2,
                "borderWidth": 1,
                "flexDirection": "row",
              },
              false,
              Object {
                "borderColor": undefined,
              },
              Object {
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "width": undefined,
              },
            ]
          }
        >
          <TextInput
            editable={false}
            onBlur={[Function]}
            onFocus={[Function]}
            placeholder="placeholder"
            placeholderTextColor="#D7D7D7"
            style={
              Object {
                "color": "#000",
                "flex": 1,
                "height": 58.971428571428575,
                "paddingLeft": 18.428571428571427,
                "textAlign": "left",
              }
            }
            value=""
          />
          <View
            style={
              Object {
                "paddingLeft": 6.142857142857142,
                "paddingRight": 18.428571428571427,
              }
            }
          >
            <SvgMock
              height={31.428571428571427}
              name="downarrow"
              width={31.428571428571427}
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "marginLeft": 0,
            "marginRight": 0,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;

exports[`renders correctly with default props 1`] = `
<View
  renderToHardwareTextureAndroid={true}
  style={
    Object {
      "marginBottom": 0,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 0,
    }
  }
>
  <View>
    <View>
      <View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFF",
                "borderRadius": 2,
                "borderWidth": 1,
                "flexDirection": "row",
              },
              false,
              Object {
                "borderColor": undefined,
              },
              Object {
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "width": undefined,
              },
            ]
          }
        >
          <TextInput
            editable={false}
            onBlur={[Function]}
            onFocus={[Function]}
            placeholder="placeholder"
            placeholderTextColor="#D7D7D7"
            style={
              Object {
                "color": "#000",
                "flex": 1,
                "height": 58.971428571428575,
                "paddingLeft": 18.428571428571427,
                "textAlign": "left",
              }
            }
            value=""
          />
          <View
            style={
              Object {
                "paddingLeft": 6.142857142857142,
                "paddingRight": 18.428571428571427,
              }
            }
          >
            <SvgMock
              height={31.428571428571427}
              name="downarrow"
              width={31.428571428571427}
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "marginLeft": 0,
            "marginRight": 0,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;

exports[`renders correctly with disabled & custom height 1`] = `
<View
  renderToHardwareTextureAndroid={true}
  style={
    Object {
      "marginBottom": 0,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 0,
    }
  }
>
  <View>
    <View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#D7D7D7",
              "borderRadius": 2,
              "flexDirection": "row",
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "width": undefined,
            }
          }
        >
          <TextInput
            editable={false}
            placeholder="placeholder"
            placeholderTextColor="#D7D7D7"
            style={
              Object {
                "flex": 1,
                "height": 58.971428571428575,
                "paddingLeft": 18.428571428571427,
                "textAlign": "left",
              }
            }
          />
          <View
            style={
              Object {
                "paddingLeft": 6.142857142857142,
                "paddingRight": 18.428571428571427,
              }
            }
          >
            <SvgMock
              height={20}
              name="downarrow"
              width={20}
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "marginLeft": 0,
            "marginRight": 0,
          }
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": "100%",
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;
